device/sycl/src/sanity/contiguous_on.hpp

```diff
@@ -26,122 +23,155 @@
+
+    // Exit early for empty containers.
+    if (n == 0) {
+        return true;
+    }
 
     // Get the output type of the projection.
-    using projection_t = std::invoke_result_t<P, T>;
+    using projection_t =
+        std::invoke_result_t<P, decltype(std::declval<CONTAINER>().at(0))>;
 
     // Allocate memory for intermediate values and outputs, then set them up.
-    vecmem::unique_alloc_ptr<projection_t[]> iout =
-        vecmem::make_unique_alloc<projection_t[]>(mr, n);
-    vecmem::unique_alloc_ptr<uint32_t> iout_size =
-        vecmem::make_unique_alloc<uint32_t>(mr);
+    vecmem::data::vector_buffer<projection_t> iout(
+        n, mr, vecmem::data::buffer_type::resizable);
+    copy.setup(iout)->wait();
     vecmem::unique_alloc_ptr<bool> out = vecmem::make_unique_alloc<bool>(mr);
 
-    uint32_t initial_iout_size = 0;
     bool initial_out = true;
 
-    cl::sycl::event kernel1_memcpy1_evt =
-        queue.copy(&initial_iout_size, iout_size.get(), 1);
-    cl::sycl::event kernel2_memcpy1_evt =
-        queue.copy(&initial_out, out.get(), 1);
+    cl::sycl::event kernel2_memcpy_evt = queue.copy(&initial_out, out.get(), 1);
 
     cl::sycl::nd_range<1> compress_adjacent_range{
-        cl::sycl::range<1>(((n + block_size - 1) / block_size) * block_size),
-        cl::sycl::range<1>(block_size)};
+        cl::sycl::range<1>(((n + local_size - 1) / local_size) * local_size),
+        cl::sycl::range<1>(local_size)};
 
     // Launch the first kernel, which will squash consecutive equal elements
     // into one element.
-    cl::sycl::event kernel1_evt = queue.submit([&](cl::sycl::handler& h) {
-        h.depends_on(kernel1_memcpy1_evt);
-        h.parallel_for<
-            kernels::IsContiguousOnCompressAdjacent<P, T, projection_t>>(
-            compress_adjacent_range,
-            [vector, projection, out = iout.get(),
-             out_size = iout_size.get()](cl::sycl::nd_item<1> item) {
-                std::size_t tid = item.get_global_linear_id();
-
-                vecmem::device_vector<T> in(vector);
-                vecmem::device_atomic_ref<uint32_t> out_siz_atm(*out_size);
-
-                if (tid > 0 && tid < in.size()) {
-                    std::invoke_result_t<P, T> v1 = projection(in.at(tid - 1));
-                    std::invoke_result_t<P, T> v2 = projection(in.at(tid));
-
-                    if (v1 != v2) {
-                        out[out_siz_atm.fetch_add(1)] = v2;
-                    }
-                } else if (tid == 0) {
-                    out[out_siz_atm.fetch_add(1)] = projection(in.at(tid));
-                }
-            });
-    });
-
-    // Copy the total number of squashed elements, e.g. the size of the
-    // resulting vector.
-    uint32_t host_iout_size;
-
     queue
-        .memcpy(&host_iout_size, iout_size.get(), sizeof(uint32_t),
-                {kernel1_evt})
+        .submit([&](cl::sycl::handler& h) {
+            h.parallel_for<kernels::is_contiguous_on_compress_adjacent<
```

Comment:
General comment about these SYCL kernels, I notice you've removed a lot of the dependencies between events. Did you ensure the sequencing of computations is correct?

Reply:
I did. I had to, because vecmem copies do not return `sycl::event` objects that we could use like this.

It did make this code a little less efficient. But since it's only used for debugging, I was not too worried.

Reply:
Okay, as long as it is still correct I am happy!